third_party/icu/googleurl/googleurl_test.cc

```diff
@@ -0,0 +1,46 @@
+#include "url/gurl.h"
+
+#include "gtest/gtest.h"
+
+namespace {
+
+void expectInvalidUrl(const std::string& input) {
+  GURL url(input);
+  EXPECT_FALSE(url.is_valid());
+}
+
+// Verifies that valid URLs are parsed correctly.
+TEST(GoogleUrl, ValidUrl) {
+  GURL url("https://example.org/test?foo=bar#section");
+  EXPECT_TRUE(url.is_valid());
+  EXPECT_EQ(url.scheme(), "https");
+  EXPECT_EQ(url.host(), "example.org");
+  EXPECT_EQ(url.EffectiveIntPort(), 443);
+  EXPECT_EQ(url.path(), "/test");
+  EXPECT_EQ(url.query(), "foo=bar");
+  EXPECT_EQ(url.ref(), "section");
+
+  GURL punycode("https://xn--c1yn36f.example");
+  EXPECT_TRUE(punycode.is_valid());
+
+  GURL percent_encoded_valid("https://%20.example");
+  EXPECT_TRUE(percent_encoded_valid.is_valid());
```

Comment:
I think this should be invalid. I think Envoy should expect that the hostname part is DNS compliant.

Reply:
Seems like we can't rely on GURL for URL parsing then?

Reply:
I do not think we can use the top level GURL class. This class does two things. It breaks URL into its constituents (scheme, userinfo, host, path, query, etc) and then it converts each constituent into its canonical form. Currently URL parser from http-parser only does the first thing - breaking up the URL. It does not canonizalize anything. Envoy **may** canonicalize path based on the `normalize_path` HCM config option using forked chrome URL code.
I think the value of bringing in chrome URL is to use the same library for both breaking UP the URL and path canonicalization. For this to work we will need to write our own `EnvoyURL` class which will use `ParseStandardURL` function from Chrome URL, which covers [the following schemes](https://chromium.googlesource.com/chromium/src/+/master/url/url_util.cc#34) but will not call `CanonicalizeStandardURL` afterwards.

Reply:
@yanavlasov thanks for the explanation on the plan ahead!